Add selinux support to file.managed

mchugh19 opened the pull request:
Requesting comments on this PR as there might be some outstanding issues or alternatives, including:
 - ~~set the policy so a restorecon would not change it~~ and for it to be able to recursively follow directories
 - ~~chcon is not the proper tool to change a context persistently. For a more proper and consistent change, semanage fcontext should be better~~
Now set with addition of `persist=True` on `file.get_selinux_context` execution module. Recursion not yet supported


### What does this PR do?
Add selinux functionality to file.managed
```
/tmp/selinux.test
  file.managed:
    - user: root
    - selinux:
        seuser:  system_u
        serole: object_r
        setype: system_conf_t
        seranage: s0
```

```
# salt-call --local state.sls selinux test=True
local:
----------
          ID: check_selinux
    Function: file.managed
        Name: /tmp/selinux.test
      Result: None
     Comment: File /tmp/selinux.test not updated
     Started: 10:54:39.418112
    Duration: 17.911 ms
     Changes:
              ----------
              selinux context:
                  ----------
                  New:
                      User: system_u Type: system_conf_t
                  Old:
                      User: unconfined_u Type: user_tmp_t

Summary for local
------------
Succeeded: 1 (unchanged=1, changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  17.911 ms
```

### What issues does this PR fix or reference?
#40703 and possibly #1349

### New Behavior
Uses `file.get_selinux_context` and `file.set_selinux_context` to determine if selinux updates are needed. As mentioned in a ticket #1349, set_selinux_context uses the chcon command, which might not be persistent. Feedback welcome.

### Tests written?
No

### Commits signed with GPG?
No

Commit:
partial work on selinux support

Commit:
working version

gtmanfred reviewed (commented):
Nice, this looks awesome, but does it add anything to changes when test=True is set?

mchugh19 commented on salt/states/file.py:
```diff
@@ -2670,6 +2690,72 @@ def managed(name,
+            if __opts__['test']:
+                ret['comment'] = 'File {0} selinux context to be updated'.format(name)
+                ret['result'] = None
+                ret['changes'] = {'selinux context': {'Old': selinux_change_orig,
```
Here are the reported changes with `test=True`

gtmanfred replied:
thanks!

mchugh19 commented:
I started looking into using semanage fcontext to make these changes persistent. I ended up writing a bunch of logic into `file.get_selinux_context` only to since find out that this was already done in `selinux.fcontext_add_or_delete_policy`. 

So please hold off on merging, as I'd like to enhance `file.get_selinux_context` to support calling `selinux.fcontext_add_or_delete_policy` to record these changes permanently, and will need review.

gtmanfred commented:
A test case for this that runs on centos would also be greatly appreciated.

mchugh19 commented:
It appears that this PR isn't quite right. The current selinux permissions check runs just fine if the file already exists, but crashes out if file.managed is deploying the file for the first time. In trying to figure out where to put this logic, it looks like a lot of the logic in file.managed is broken out into execution modules like `file.check_perms` which perform actions and pass back dicts including ret['changes'] which are then checked further in file.managed. 

So it looks like multiple file modules may need to be adjusted to enhance for selinux perms checking/setting. Is that accurate and the proper way forward?

Commit:
enhance file.set_selinux_context module to support calling fcontext. Use this module in file state

Commit:
support selinux attr setting for new and existing files

mchugh19 commented:
As mentioned, took a stab at support selinux attributes by way of modifying a bunch functions in the file module. 

State on new file
```
local:
----------
          ID: check_selinux
    Function: file.managed
        Name: /tmp/selinux.test
      Result: True
     Comment: Empty file; The file /tmp/selinux.test is set to be changed
     Started: 15:36:41.301914
    Duration: 1496.069 ms
     Changes:
              ----------
              new:
                  file /tmp/selinux.test created
              selinux:
                  ----------
                  New:
                      User: system_u Type: system_conf_t
                  Old:
                      User: unconfined_u Type: user_tmp_t
              user:
                  testuser2

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   1.496 s
```

Running state a second time (no changes)
```
local:
----------
          ID: check_selinux
    Function: file.managed
        Name: /tmp/selinux.test
      Result: True
     Comment: File /tmp/selinux.test exists with proper permissions. No changes made.
     Started: 15:32:35.324715
    Duration: 17.839 ms
     Changes:

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:  17.839 ms
```

Existing file needing selinux attribute updates
```
local:
----------
          ID: check_selinux
    Function: file.managed
        Name: /tmp/selinux.test
      Result: True
     Comment: The file /tmp/selinux.test is set to be changed
     Started: 15:32:29.984517
    Duration: 1530.755 ms
     Changes:
              ----------
              selinux:
                  ----------
                  New:
                      Type: system_conf_t
                  Old:
                      Type: user_tmp_t

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   1.531 s
```

As mentioned this should really get some integration tests, but that's something I've not worked on before, so I'll look that over next.

Commit:
cleanup variable assignment to pass unit tests

Commit:
fix up a couple of unrelated lint errors

Commit:
add unit tests for get and set selinux_context

Commit:
add test for check_perms with selinux modification

Commit:
merge with develop

mchugh19 commented:
@gtmanfred added a couple of unit tests for the file module. Is that the right track?
Also the file module uses functions from the selinux module. Is it correct to use `__salt__['selinux` or should there be an import?

gtmanfred commented:
That is a great start, but i would really like to see a test for the `file.managed` addition, to test applying selinux roles to a file managed with file.managed.

thanks,
Daniel

mchugh19 commented:
@gtmanfred The `file.managed` state ends up using the `file.check_perms`, `file.check_managed_changes`, and `file.manage_file` execution modules for all logic. So that's why I went with the `file.check_perms` unit test. Happy to work on the `file.managed` state test next. 
Is it acceptable to have the state unit test actually run on a file, or is best practice to mock all internal components? The `file.managed` state is very complex, so I'd be waaaay happier about actually running it ;)

Also the file module uses functions from the selinux module. Is it correct to use __salt__['selinux or should there be an import?

gtmanfred commented:
If you want it to touch an actual file, it should probably be an integration test, and not a unit test.

If it is a unit test, you should mock all the components.

You should never import modules, it should always be called from the loader object, so selinux functions from `__salt__`

Commit:
Add file.managed selinux state test
Add additional file module selinux test
Remove extra spaces in file module change output
Better support setting selinux context other than type
Cleanup of some file module tests which leave around files in /tmp (there are still many more)

mchugh19 commented:
State test added for file.manage when passing selinux user and type. Is there anything else outstanding?

Commit:
correct check_file_meta output

gtmanfred reviewed (approved):
This looks great thanks!

Commit:
Merge branch 'develop' into 40703

terminalmage commented on salt/modules/file.py:
```diff
@@ -3841,10 +3842,23 @@ def set_selinux_context(path,
+
+    .. versionchanged:: Neon
+
+        Added persist option
```
Sorry for this nitpick, but we try to put our versionchanged/versionadded notices at the top of the docstring rather than the bottom. Could you move this up?

mchugh19 replied:
No problem! Done!

terminalmage commented on salt/modules/file.py:
```diff
@@ -4485,6 +4500,10 @@ def check_perms(name, ret, user, group, mode, attrs=None, follow_symlinks=False)
+
+    .. versionchanged:: Neon
+
+        Added selinux options
```
Same here, this needs to go to the top of the docstring.

mchugh19 replied:
... and also moved! Blamity!

terminalmage commented on salt/modules/file.py:
```diff
@@ -4868,6 +4997,26 @@ def check_file_meta(
+    serange
+        selinux range attribute
+
+        .. versionadded:: Neon
```
These are OK since they denote individual arguments :+1: 

It's just the ones that apply to the entire function that need to be at the top.

terminalmage reviewed (changes requested):
The PR looks great, there are just two changes in the docstrings that I would like to see be made for consistency with the rest of our docs.

Commit:
move versionchanged to proper spot

Commit:
Merge branch '40703' of github.com:mchugh19/salt into 40703

rallytime commented:
@mchugh19 There are some lint issues that need to be addressed here as well. Can you take another look?

Commit:
correct lint errors

mchugh19 commented:
I think I got all the lint errors

gtmanfred committed:
Merge branch 'develop' into 40703

cachedout commented:
@mchugh19 Looks like we have an unhappy test: https://jenkinsci.saltstack.com/job/pr-kitchen-centos7-py3/job/PR-49860/21/testReport/junit/unit.modules.test_file/FileSelinuxTestCase/test_file_check_perms/

Commit:
Merge branch 'develop' into 40703

Commit:
correct test to remove extra space

Commit:
Merge branch '40703' of github.com:mchugh19/salt into 40703

gtmanfred commented:
:+1: totally worth adding a not in the neon release notes.

Commit:
update release notes with new selinux functionality

mchugh19 commented:
Release notes note added

Commit:
Merge branch 'develop' into 40703

rallytime commented:
Hi @mchugh19 - The centos tests are failing a test here now. Can you take another look?

https://jenkinsci.saltstack.com/job/pr-kitchen-centos7-py2/job/PR-49860/26/

Commit:
disable lsattr calls on selinux test

Commit:
Merge branch '40703' of github.com:mchugh19/salt into 40703

mchugh19 commented:
Corrected selinux test to disable lsattr calls. Looks good now!
